src/control/JacobianController.h

```diff
@@ -0,0 +1,139 @@
+ * @see https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant
+ */
+template <int outputDim, int inputDim> class JacobianController {
+public:
+	/**
+	 * @brief Construct a new controller.
+	 *
+	 * @param kinematicsFunc The kinematics of the mechanism being controlled. This is
+	 * a function that takes in some input vector (ex: joint angles of an arm) and outputs
+	 * another vector (ex: the 3d pose of the hand)
+	 * @param jacobianFunc The jacobian of the kinematic function.
+	 * When executing a command, getCommand() should be called at this rate.
+	 * @param maxVels The maximum velocity in the output space of the kinematic function.
+	 * @param maxAccels The maximum acceleration in the output space of the kinematic function.
+	 */
+	JacobianController(
+		const std::function<navtypes::Vectord<outputDim>(const navtypes::Vectord<inputDim>&)>&
+			kinematicsFunc,
+		const std::function<navtypes::Matrixd<outputDim, inputDim>(
+			const navtypes::Vectord<inputDim>&)>& jacobianFunc,
+		double maxVel, double maxAccel)
+		: kinematicsFunc(kinematicsFunc),
+		  jacobianFunc(jacobianFunc
+						   ? jacobianFunc
+						   : std::bind(filters::statespace::numericalJacobian, kinematicsFunc,
+									   std::placeholders::_1, outputDim)),
+		  velocityProfile(maxVel, maxAccel) {
+		assert(this->kinematicsFunc);
+		assert(this->jacobianFunc);
+	}
+
+	/**
+	 * @brief Get the current command.
+	 *
+	 * @param currTime The current timestamp.
+	 * @param currPos The current position, in the input space.
+	 * @return navtypes::Vectord<inputDim> The target position, in the input space of the
+	 * kinematic function. Returns the current position if no target is set.
+	 */
+	navtypes::Vectord<inputDim> getCommand(robot::types::datatime_t currTime,
+										   const navtypes::Vectord<inputDim>& currPos) const {
+		double unused;
+		return getCommand(currTime, currPos, unused);
+	}
+
+	/**
+	 * @brief Get the current command.
+	 *
+	 * @param currTime The current timestamp.
+	 * @param currPos The current position, in the input space.
+	 * @param[out] cosineSim Output parameter, gives the cosine similarity of the delta to the
+	 * returned target and the delta to the commanded target. This similarity may be low if the
+	 * kinematics of the mechanism do not allow it to move in the commanded direction. May be
+	 * NaN if stuck in singularity point, in which case the command is the current position.
+	 * @return navtypes::Vectord<inputDim> The target position, in the input space of the
+	 * kinematic function. Returns the current position if no target is set.
+	 */
+	navtypes::Vectord<inputDim> getCommand(robot::types::datatime_t currTime,
+										   const navtypes::Vectord<inputDim>& currPos,
+										   double& cosineSim) const {
+		if (!velocityProfile.hasTarget()) {
+			return currPos;
+		}
+
+		navtypes::Vectord<outputDim> currTarget = velocityProfile.getCommand(currTime);
```

Comment:
I think it would be good to add comments explaining what is happening at a high level.